x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Realm.java

```diff
@@ -144,17 +157,45 @@ public void usageStats(ActionListener<Map<String, Object>> listener) {
         listener.onResponse(stats);
     }
 
+    public RealmRef getRealmRef() {
+        return realmRef;
+    }
+
+    public Set<RealmRef> getDomainRealmRef() {
+        if (null == domainRealmRef) {
+            throw new IllegalStateException("Realm has not been initialized");
+        }
+        return domainRealmRef;
+    }
+
     @Override
     public String toString() {
-        return config.type() + "/" + config.name();
+        if (domain() != null) {
+            return config.type() + "/" + config.name() + "/" + config.domain();
+        } else {
+            return config.type() + "/" + config.name();
+        }
     }
 
     /**
-     * This is no-op in the base class, but allows realms to be aware of what other realms are configured
+     * This allows realms to be aware of what other realms are configured
      *
      * @see DelegatedAuthorizationSettings
      */
-    public void initialize(Iterable<Realm> realms, XPackLicenseState licenseState) {}
+    public void initialize(Iterable<Realm> realms, XPackLicenseState licenseState) {
+        final String domainName = getRealmRef().getDomain();
+        if (null == domainName) {
+            this.domainRealmRef = Set.of();
+            return;
+        }
+        final Set<RealmRef> domainRealmRefBuilder = new HashSet<>();
+        for (Realm otherRealm : realms) {
+            if (false == getRealmRef().equals(otherRealm.getRealmRef()) && domainName.equals(otherRealm.getRealmRef().getDomain())) {
+                domainRealmRefBuilder.add(otherRealm.getRealmRef());
+            }
+        }
```

Comment:
It may not be possible to have phantom realms with this approach. Do we care?

Reply:
Phantom realms are not supported, also because realm settings validations in other places, from a previous PR.
We discussed that we're not going to tackle phantom realms for now, so I don't think we should code for it here, in this PR.